tests/e2e.spec.ts

```diff
@@ -0,0 +1,282 @@
+/**
+ * @license
+ * Copyright 2021 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect } from 'chai';
+import { randomBytes } from 'crypto';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+
+import { read, write } from '../build/config';
+import { getBootPartition } from '../build/utils';
+import { downloadOS, extractFromZipArchive, isZipFile } from './utils';
+
+import type { BalenaSDK, DeviceType } from 'balena-sdk';
+
+const fullTest = false;
+const quickTestDeviceTypes = [
+	'raspberrypi3', // very popular with partition number 1, non-flasher image
+	'intel-nuc', // very popular with partition number 1, flasher image
+	// Intel-Edison is an odd one, needs unzipping the downloaded image to extract
+	// 'resin-image-edison.hddimg', and that image contains an empty partition table!
+	'intel-edison',
+	'jetson-xavier', // largest partition number, 37
```

Comment:
These 4 device types were chosen to be representatives of all others when running automated CI tests. They run reasonably quickly, despite dynamically downloading the OS images to a temp folder.

On my laptop, with attached storage, I have also run the tests for all public device types (see full list in the `expectedConfigBySlug` array in this file), which took hours and 120GB of disk space.

In addition to the automated tests on Linux, I have also tested manually alongside CLI PR balena-io/balena-cli/issues/2349, on macOS and Windows.